File: client/src/types/job.ts
/**
 * Domain types describing a conversion job and its lifecycle.
 * These mirror the shapes returned by the backend API.
 */

/** Unique identifier returned by the backend after an upload. */
export type JobId = string;

/**
 * The lifecycle states a conversion job moves through.
 * Ordered roughly by progression, but the backend is the source of truth.
 */
export type JobStatus =
  | "uploaded"
  | "generating_slides"
  | "generating_audio"
  | "rendering_video"
  | "completed"
  | "failed";

/** Response from POST /upload */
export interface UploadResponse {
  job_id: JobId;
}

/** A single slide derived from the markdown headings. */
export interface Slide {
  /** Stable identifier for the slide (used as a React key). */
  id: string;
  /** Slide title, typically derived from a markdown heading. */
  title: string;
  /** 1-based position of the slide within the deck. */
  index: number;
  /** Optional URL to a rendered thumbnail image. */
  thumbnail_url?: string;
}

/** Response from GET /slides/{job_id} */
export interface SlidesResponse {
  job_id: JobId;
  slides: Slide[];
}

/** Response from GET /status/{job_id} */
export interface StatusResponse {
  job_id: JobId;
  status: JobStatus;
  /** Optional 0-100 progress percentage for the current stage. */
  progress?: number;
  /** Human-readable message, present especially when status is "failed". */
  message?: string;
}

/** Response from GET /result/{job_id} */
export interface ResultResponse {
  job_id: JobId;
  /** URL of the rendered video, playable in an HTML5 <video> element. */
  video_url: string;
  /** Optional direct download URL; falls back to video_url if absent. */
  download_url?: string;
  /** Optional duration of the rendered video in seconds. */
  duration_seconds?: number;
}

/** Response from POST /generate/{job_id} */
export interface GenerateResponse {
  job_id: JobId;
  status: JobStatus;
}

File: server/app/schemas/job.py
"""Request/response schemas for the conversion API.

These mirror the documented API contract. A few extra, harmless fields
(``index``, ``thumbnail_url``, ``job_id``, ``status``) are included so the
existing frontend can consume the responses without changes.
"""

from __future__ import annotations

from pydantic import BaseModel, Field

from app.models.job import JobStatus


class UploadResponse(BaseModel):
    """Response for POST /upload."""

    job_id: str = Field(..., description="Unique identifier for the job.")
    filename: str = Field(..., description="Original uploaded filename.")


class SlideItem(BaseModel):
    """A single generated slide preview."""

    id: int = Field(..., description="1-based slide id.")
    title: str = Field(..., description="Slide title from the heading.")
    image_url: str = Field(..., description="Public URL of the slide image.")
    # Aliases for frontend compatibility:
    index: int = Field(..., description="Same as id; 1-based position.")
    thumbnail_url: str = Field(..., description="Same as image_url.")


class SlidesResponse(BaseModel):
    """Response for GET /slides/{job_id}."""

    job_id: str
    slides: list[SlideItem]


class GenerateResponse(BaseModel):
    """Response for POST /generate/{job_id}."""

    message: str
    job_id: str
    status: JobStatus


class StatusResponse(BaseModel):
    """Response for GET /status/{job_id}."""

    job_id: str
    status: JobStatus
    message: str | None = Field(
        default=None, description="Error or informational message."
    )


class VideoResultResponse(BaseModel):
    """Response for GET /result/{job_id}."""

    job_id: str
    video_url: str
    download_url: str
